src/wallet/rpcdump.cpp

```diff
@@ -892,103 +886,95 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
             if (!pwallet->HaveCScript(redeem_id) && !pwallet->AddCScript(redeemScript)) {
                 throw JSONRPCError(RPC_WALLET_ERROR, "Error adding p2sh redeemScript to wallet");
             }
 
-            CScript redeemDestination = GetScriptForDestination(redeem_id);
+            // Check for P2SH-P2WSH
+            if (redeemScript.IsPayToWitnessScriptHash()) {
+                if (!IsHex(witness_script_hex)) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid witness script");
+                }
 
-            if (::IsMine(*pwallet, redeemDestination) == ISMINE_SPENDABLE) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this address or script");
-            }
+                // Generate the scripts
+                std::vector<unsigned char> witness_script_parsed(ParseHex(witness_script_hex));
+                CScript witness_script = CScript(witness_script_parsed.begin(), witness_script_parsed.end());
+                CScriptID witness_id(witness_script);
 
-            pwallet->MarkDirty();
+                // Check that the witnessScript and P2SH redeemScript match
+                if (GetScriptForDestination(WitnessV0ScriptHash(witness_script)) != redeemScript) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "The witnessScript does not match the redeemScript");
+                }
 
-            if (!pwallet->AddWatchOnly(redeemDestination, timestamp)) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
-            }
+                // Import into the wallet
+                if (!pwallet->AddWatchOnly(witness_script, timestamp)) {
+                    throw JSONRPCError(RPC_WALLET_ERROR, "Error adding address to wallet");
+                }
 
-            // add to address book or update label
-            if (IsValidDestination(dest)) {
-                pwallet->SetAddressBook(dest, label, "receive");
+                if (!pwallet->HaveCScript(witness_id) && !pwallet->AddCScript(witness_script)) {
+                    throw JSONRPCError(RPC_WALLET_ERROR, "Error adding p2sh-p2wsh witnessScript to wallet");
+                }
             }
 
-            // Import private keys.
-            if (keys.size()) {
-                for (size_t i = 0; i < keys.size(); i++) {
-                    const std::string& privkey = keys[i].get_str();
-
-                    CKey key = DecodeSecret(privkey);
-
-                    if (!key.IsValid()) {
-                        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid private key encoding");
-                    }
-
-                    CPubKey pubkey = key.GetPubKey();
-                    assert(key.VerifyPubKey(pubkey));
-
-                    CKeyID vchAddress = pubkey.GetID();
-                    pwallet->MarkDirty();
-                    pwallet->SetAddressBook(vchAddress, label, "receive");
+        // P2WSH
```

Comment:
// (P2SH-)P2WSH

Makes it more clear to me that the above block changed the `script`

Reply:
And same thing with: `// P2PK/P2PKH/P2WPKH`